src/types/web-resource.ts

```diff
@@ -0,0 +1,108 @@
+import * as TypeUtils from '../type-utils';
+import { storageRegistry } from '../storage-adapters/index';
+
+/**
+ * This is how WebResources are stored on the DB
+ */
+export type WebResourceRef = {
+	filename: string;
+	href: string;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+};
+
+/**
+ * This is how WebResources are defined on creation
+ */
+export type WebResourceInput = {
+	filename: string;
+	data: Buffer;
+	contentType?: string;
+	contentDisposition?: string;
+	size?: number;
+	storage: string;
+};
+
+export const types = {
+	postgres: 'JSONB',
+	mysql: 'JSON',
+	websql: 'TEXT',
+	odata: {
+		name: 'Self.WebResource',
+		complexType: `\
+<ComplexType Name="WebResource">
+	<Property Name="filename" Nullable="false" Type="Edm.String"/>\
+	<Property Name="href" Nullable="false" Type="Edm.String"/>\
+	<Property Name="contentType" Nullable="true" Type="Edm.String"/>\
+	<Property Name="contentDisposition" Nullable="true" Type="Edm.String"/>\
+	<Property Name="size" Nullable="true" Type="Edm.Int64"/>\
+</ComplexType>`,
+	},
+};
+
+export const fetchProcessing = (data: any) => {
+	let refData: WebResourceRef;
+	if (data === null) {
+		return data;
+	}
+	if (typeof data === 'string') {
+		refData = JSON.parse(data);
+	} else if (typeof data === 'object') {
+		refData = data;
+	} else {
+		throw new Error(
+			'Unable to parse data for a WebResource type. Data is: ' + data,
+		);
+	}
+
+	return {
+		filename: refData.filename,
+		href: refData.href,
+		contentType: refData.contentType,
+		contentDisposition: refData.contentDisposition,
+		size: refData.size,
+	};
+};
+
+export const validate = TypeUtils.validate.checkRequired(
+	async (value: WebResourceInput) => {
+		if (typeof value !== 'object') {
+			throw new Error('received value is not an object');
+		}
+		if (!value.filename) {
+			throw new Error('filename is required');
+		}
+		if (!value.data) {
+			throw new Error('data is required');
+		}
+
+		const storage = value.storage;
+		if (!storage) {
+			throw new Error('storage is required');
+		}
+		const storageAdapter = storageRegistry[storage];
+		if (!storageAdapter) {
+			throw new Error(
+				`storage ${storage} not defined. Available adapters: ${Object.keys(
+					storageRegistry,
+				)}`,
+			);
+		}
+
+		const webresource = await storageAdapter.saveFile(
+			value.filename,
+			value.data,
+		);
```

Comment:
What happens if the saveFile is not successfull? Will it throw an error, is the return different?

As we use this as an atomic value we may need to make sure that the atomic value is consistent. IF we want to store a new row or update a row with this webresource and the webresource fails, do we want to commit the other data? I'm not fully sure what is the error case on this webresource?

Reply:
It will throw an Error.
I've added a test and error-handling logic.

About the atomic aspect, the approach is to fail the row update/insert if we cannot save the webresource, to avoid inconsistencies